### DISCUSSION TITLE
Query: can widget visibility be improved when using dark GUI colours? [GUI Theming]

### DISCUSSION THREAD

**Discussion by giantclambake** [BlitterStudio/amiberry#1293] [General]
When using dark colored GUI themes, things can end up looking like this ;

![ex](https://imgur.com/cDKRnvm.png)

The shading works well wrt what options are active/unavailable, however it's really hard to see which options are selected.

Same image again, with widget color changed to white;

![ex2](https://imgur.com/l5pnset.png)

I do not know if the color of widgets can be changed ~ I chose to example white ..ie; simple color invert... because that would seem to me to be the quick easy fix, and stands apart from font color....I kinda like that look ;)

Or assign a color to widget background?...(white again in example, but might be user configurable);

![ex3](https://imgur.com/A8yEqOY.png) 

Is there anything along these lines that can be done to improve selection widget visibility?

TIA

(amiberry.conf color settings used above)

```
gui_theme_font_color=17, 243, 50
gui_theme_base_color=54, 8, 181
gui_theme_selector_inactive=53, 7, 87
gui_theme_selector_active=43, 88, 120
gui_theme_selection_color=15, 27, 217
gui_theme_textbox_background=44, 42, 43
```

**Comment by giantclambake** [comment]
...yeah, this is basically the same situation, that spawn the need for configurable font color ... unfortunately speaking, I'd erroneously thought the 'dot' in radioboxes and 'tick' in checkboxes, would inherit the value of font color ; my bad  ;^/

I cannot see anywhere in libguisan/sdl that allows for color selection of these widget characters ...I very easily could've missed something along the way, but what I do note is a 'technical misnomer'....ie;

widget background color is set to  colTextboxBackground as exposed by gui_theme_textbox_background

Apparently, the problem I have with that is these are widgets, not textboxs, and something like colWidgetBackground should exist, as exposed by a hypothetical gui_theme_widget_background variable .... that would be more 'correct' as it were.

However .. the background color isn't really the crux problem ~ it's the fact the 'dot' & 'tick' are always rendered black ...

footnote: the little arrows on dropdown selectors aren't really a problem, as it's (visually) evident when a dropdown menu is active or not, and the current selection is clearly displayed... as these are textboxs and font color applies..

**Comment by giantclambake** [comment]
...and yes, I did notice the Savestate panel looks like this...

![ex](https://imgur.com/yOVRUwk.png)

This is because those radiobuttons do not have colTextboxBackground set  (somehow they got missed)

That should actually be fixed for sake of consistency, but as it's part of this same entanglement, it's discussion .. not a bug ;)

**Reply by midwan** [reply]
Fixed with cda18ef0aa1ad553bf320f75dae9a9508f297efd

**Reply by giantclambake** [reply]
Close ....but you'd never pick this up using default colors (luckily I have a penchant for dark themes ;)

![ex](https://imgur.com/cKd3N82.png)

...looks like the radioboxes have a 1 pixel white border, whereas all the other GUI radiobox widgets do not?

TIA

**Reply by midwan** [reply]
Fixed with eed7f2833d8ab5fc569a686faafe256d3353138f

**Reply by giantclambake** [reply]
Better ;)

**Comment by giantclambake** [comment]
If you're after consistency of widgets across the GUI, then the following checkboxes need attention...

Paths panel...(logging checkboxes)

![ex](https://imgur.com/elynBmV.png)


Display panel... (Manual crop checkbox)

![ex2](https://imgur.com/YNskeq6.png)


WHDLoad panel ... (Slave Libraries checkbox)

![ex3](https://imgur.com/iF9kW0M.png)


Once those are addressed, widget colors are consistent across all GUI panels ;)

**Reply by midwan** [reply]
should be fixed now :)

**Reply by giantclambake** [reply]
Groovy, all ducks in a row now =)

**Comment by giantclambake** [comment]
@midwan  ~ thanks for all these small fixups ; fwiw, these commits have formed a set of 'instructional examples' for me, to better understand how libguisan works, and how amiberry implements it.... I may not be able to code, but I can read ;)... [glimpse](https://github.com/gvelez17/glimpse) is my long time weapon of choice... 

The commit above (eed7f28) demonstrated to me that if `setBaseColor(gui_baseCol)` is not used, the widgets fall thru to (I think?) base window color (actually mBackgroundColor(0xffffff)  ;). I went back and read the libguisan widget stuffs closely, then compared that read against the amiberry panel source files wrt widget creation, and the one thing they shared in common, was that `setForegroundColor`  was unused. This inferred to me that the widgets must also be falling thru to the base definition of ForegroundColor, which is;

`external/libguisan/src/widget.cpp: 		: mForegroundColor(0x000000),`

Armed with these observations, I loaded a dark theme into xfce4 WM (darkened window title bar, window title font in white), and set about creating a very dark amiberry GUI theme ... it looks like this;

![ex](https://imgur.com/L1iBYAO.png)

M'kay, edit `external/libguisan/src/widget.cpp` and change to `: mForegroundColor(0xeeeeee),` and recompile ;

![ex2](https://imgur.com/EQTCKpn.png)

Huzzah! So it is possible, without contemplating the horrowshow task of having to edit every piece of widget code in the GUI, because in my testing here, this can apply 'globally' to the amiberry GUI...if you want so see for yourself, here's the color settings used for the above examples (try it without/with the libguisan edit above for full effect =) ;

```
gui_theme_font_color=10, 240, 10
gui_theme_base_color=0, 0, 20
gui_theme_selector_inactive=53, 7, 87
gui_theme_selector_active=28, 28, 150
gui_theme_selection_color=15, 21, 217
gui_theme_textbox_background=44, 0, 0
```

Obviously it adds a white border around current panel selection button, but that's not necessarily a bad thing, however it does also make the dropdown menu arrow widgets visible as well. This has the effect of visually balancing those panels, that have white background slider widgets....ie; that looks a little garish with black widget foreground, however panel looks ok with white widget foreground with white slider background (when active).

In winuae you have a 'dark mode' widget in the GUI ~ it wouldn't be too far out of step, for amiberry.conf to have an analog setting of ..ie; `gui_dark_mode=<boolean>` ...even if in effect it only changes the mForegroundColor value from 000000 to eeeeee  ;)

As this is only impacting widgets and things that should be seen in the darkness, I think a choice between black/white is really all that's required here, personally speaking (like the various dark themes for xfce4, most simply toggle what was black, to white... like window titlebar text).

Is it at all possible to implement this option in amiberry?

TIA

**Reply by midwan** [reply]
I'm afraid this still needs to be added as an extra field, and assigned to all the widgets. The library is supposed to be generic, we shouldn't be hardcoding stuff in there :)

**Reply by midwan** [reply]
@giantclambake OK, `preview` now has support for foreground color in the GUI theme. ;)

**Reply by giantclambake** [reply]
@midwan Outstanding! I was so hoping you could avoid the edits, but I get where you're coming from. Also noted you addressed slider background color as well... thanks, looks much better now with dark theming =)

I'll carefully go through the GUI panels again, to make sure there's no more 'stray puppies' ;)

**Comment by giantclambake** [comment]
Excellent! I believe this wraps up all the lose ends wrt GUI theming once and for all =)

Closing as fixed by various commits here-above.
